apps/prairielearn/src/tests/workspaceDynamicFiles.test.ts

```diff
@@ -2,175 +2,202 @@ import * as fs from 'node:fs/promises';
-        { file: 'Dynamic file 1', msg: 'does not include a name' },
-        { file: 'invalid_encoding.bin', msg: 'unsupported file encoding' },
-        { file: '../outside_home.txt', msg: 'traverses outside the home directory' },
-        { file: 'path/../../outside_home.txt', msg: 'traverses outside the home directory' },
-        { file: '/home/prairie/absolute.txt', msg: 'has an absolute path' },
-        { file: 'server.py', msg: 'local file outside the question directory' },
-        { file: 'no_contents.txt', msg: 'has neither "contents" nor "questionFile"', contents: '' },
-      ];
-      for (const expectedError of expectedErrors) {
-        const issueError = issueErrors.find((error) => error.file === expectedError.file);
-        assert.isDefined(issueError, `Expected error not found: ${expectedError.file}`);
-        assert.include(issueError.msg, expectedError.msg);
-        if (expectedError.contents != null) {
-          await checkFileContents(expectedError.file, expectedError.contents);
-        } else {
-          await fs.access(join(sourcePath, expectedError.file)).then(
-            () => assert.fail(`File ${expectedError.file} should not exist`),
-            (err) => assert.equal(err.code, 'ENOENT'),
-          );
-        }
+    assert.lengthOf(fileGenerationErrors, 7);
+
+    // Valid files should still have been created.
+    await checkFileContents(targetPath.path, 'static.txt', 'Static content\n');
+    await checkFileContents(targetPath.path, 'template.txt', 'A is STRING, while B is 53\n');
+    await checkFileContents(targetPath.path, 'dynamic.txt', 'This is a dynamic file.\n');
+
+    const expectedErrors = [
+      { file: 'Dynamic file 1', msg: 'does not include a name' },
+      { file: 'invalid_encoding.bin', msg: 'unsupported file encoding' },
+      { file: '../outside_home.txt', msg: 'traverses outside the home directory' },
+      { file: 'path/../../outside_home.txt', msg: 'traverses outside the home directory' },
+      { file: '/home/prairie/absolute.txt', msg: 'has an absolute path' },
+      { file: 'server.py', msg: 'local file outside the question directory' },
+      { file: 'no_contents.txt', msg: 'has neither "contents" nor "questionFile"', contents: '' },
+    ];
+
+    for (const expectedError of expectedErrors) {
+      const issueError = fileGenerationErrors.find((error) => error.file === expectedError.file);
+      assert.isDefined(issueError, `Expected error not found: ${expectedError.file}`);
+      assert.include(issueError.msg, expectedError.msg);
+      if (expectedError.contents != null) {
+        await checkFileContents(targetPath.path, expectedError.file, expectedError.contents);
+      } else {
+        await fs.access(join(targetPath.path, expectedError.file)).then(
+          () => assert.fail(`File ${expectedError.file} should not exist`),
+          (err) => assert.equal(err.code, 'ENOENT'),
+        );
       }
-      assert.lengthOf(issueErrors, expectedErrors.length);
```

Comment:
Any reason why you moved the length to before the loop? This was moved to after the loop mainly to make sure that, if the test doesn't pass because an error that was supposed to be caught wasn't caught, the test shows that assertion error first instead of the count, making debugging the test easier. So mocha will tell me that the error is in "server.py", instead of telling me that 7 errors were expected but 6 were found.

Reply:
I can reorder it 👍

Reply:
Reordered in 744098d.